### PR TITLE
[WIP] Support exists

### LINKED ISSUE #426: Support exists query
**Summary**

```
select number from numbers_mt(10) where number exists (select number from numbers_mt(1));
```

Need by TPCH q4:
```
select
    o_orderpriority,
    count(*) as order_count
from
    orders
where
        o_orderdate >= '1993-07-01'
  and o_orderdate < date '1993-07-01' + interval '3' month
  and exists (
        select
            *
        from
            lineitem
        where
                l_orderkey = o_orderkey
          and l_commitdate < l_receiptdate
    )
group by
    o_orderpriority
order by
    o_orderpriority;
```

### PR DESCRIPTION
**Pull request by jyizheng**
I hereby agree to the terms of the CLA available at: https://datafuse.rs/policies/cla/

## Summary

support exists  in where expression

## Changelog

- New Feature

## Related Issues

Fixes #426 

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

**Comment by sundy-li** [comment]
https://dev.mysql.com/doc/refman/8.0/en/exists-and-not-exists-subqueries.html

File: fusequery/query/src/datasources/system/numbers_stream.rs [hunk 1]
```diff
@@ -52,11 +52,15 @@ impl NumbersStream {
...
                 return Ok(None);
             }
+            if partitions.len() == 1 && partitions[0].name.is_empty() {
+                return Ok(None);
+            }
 
             let block_size = self.ctx.get_max_block_size()?;
             let mut blocks = Vec::with_capacity(partitions.len());
-
+            println!("partitions={:?}", partitions);
```
Comment (sundy-li) [inline fusequery/query/src/datasources/system/numbers_stream.rs:61]: unused logs

File: fusequery/query/src/datasources/system/numbers_stream.rs [hunk 2]
```diff
@@ -52,11 +52,15 @@ impl NumbersStream {
...
             for part in partitions {
+                println!("names={:?}", part.name);
```
Comment (sundy-li) [inline fusequery/query/src/datasources/system/numbers_stream.rs:63]: .

File: fusequery/query/src/pipelines/transforms/transform_expression_executor.rs [hunk 3]
```diff
@@ -51,7 +52,11 @@ impl ExpressionExecutor {
         Ok(())
     }
 
-    pub fn execute(&self, block: &DataBlock) -> Result<DataBlock> {
+    pub fn execute(
+        &self,
+        block: &DataBlock,
+        exists_res: Option<&HashMap<String, bool>>,
```
Comment (sundy-li) [inline fusequery/query/src/pipelines/transforms/transform_expression_executor.rs:58]: Introduce `HashMap<String, bool> ` is too limited just for exists sub query,  I think we should introduce a new struct to store the subquery results. Such as 

```
struct SubqueryResult {
       blocks: Map<String, DataBlock>
}
```

File: fusequery/query/src/interpreters/interpreter_select.rs [hunk 4]
```diff
@@ -43,36 +105,55 @@ impl IInterpreter for SelectInterpreter {
...
-                    // TODO: ISSUE-204 kill prepared query stage
-                    killed_set.insert(node.name.clone());
+        while queue1.len() > 0 {
+            let mut one_level = Vec::<PlanNode>::new();
+            while queue1.len() > 0 {
+                if let Some(begin) = queue1.pop_front() {
+                    if let Ok(p) = get_filter_plan(begin) {
+                        let exists_vec = find_exists_exprs(&[p.predicate.clone()]);
+                        for exst in exists_vec {
+                            let expr_name = format!("{:?}", exst);
```
Comment (sundy-li) [inline fusequery/query/src/interpreters/interpreter_select.rs:126]: we have `expr.column_name()` to identify an expr.

File: fusequery/query/src/pipelines/transforms/transform_expression_executor.rs (hunk 3 above)
Comment (jyizheng) [reply fusequery/query/src/pipelines/transforms/transform_expression_executor.rs:58]: I see, this is much cleaner. I just tried this approach and it turns out `DataBlock` is not serializable or deserializable.
To fix this, I need to add  `#[derive(serde::Serialize, serde::Deserialize)]` to `DataBlock` and `DataColumnarValue` and `DataArrayRef`. 

However, `DataArrayRef` is imported from arrow. I am not sure how to add the de/serialization trait for them.
Is there any workaround?

**Comment by jyizheng** [comment]
@sundy-li @zhang2014 

Initially, we planned to send DataStream to the flight server so that Exists and IN can be implemented similarly.
It seems the current code does not support this approach very well---it is not very convenient to send DataStream to the flight server. 

Would it be okay to merge this PR for now?
I will keep improving the implementation of Exists and exploring the approach of implementation of IN operator in future PRs.

**Comment by jyizheng** [comment]
@sundy-li  @zhang2014 
PTAL

**Review by sundy-li — APPROVED** [review]

**Comment by BohuTANG** [comment]
Realy cool, thank you @jyizheng

**Pull request merged**